behavior/plans/diff.py

```diff
@@ -1,466 +1,288 @@
+    end_matches = [np.argwhere(root_end_times >= end) for end in end_times]
+
+    # For each OU data point, find the intersection between root plan nodes that start earlier and end afterwards.
+    intersects = [
+        np.intersect1d(start_match, end_match) for (start_match, end_match) in zip(start_matches, end_matches)
+    ]
+
+    # If there is an intersection, then the OU must belong to that "invocation".
+    subinvocations[:] = [-1 if len(intersect) == 0 else intersect[0] for intersect in intersects]
+
+
+def process_query_invocation(subframe, diffed_matrixes):
+    """
+    Function used to difference all data associated with a given query session template.
+
+    Parameters
+    ----------
+    subframe : pd.DataFrame
+        Dataframe contains the data that we want to difference. The dataframe must be
+        data that is associated with a given query session template.
 
-        # Find all referenced node ids.
-        # Discard placeholder ("NULL") values for child plan node ids.
-        referenced_node_ids = set(
-            int(node_id)
-            for node_id in query_df[["left_child_plan_node_id", "right_child_plan_node_id"]].to_numpy().ravel()
-            if node_id != -1
+        In other words, <query_id, statement_timestamp, pid> is the same for all rows.
+
+    diffed_matrixes : list[np.pdarray]
+        Output list to store diffed numpy matrixes.
+    """
+    root_plans_times = subframe[subframe["plan_node_id"] == 0]
+    if root_plans_times.shape[0] > 1:
+        # This is the case where there are multiple root plan node IDs detected directly.
+        # In this case, we actually need to run separate_subinvocation to separate them.
+        #
+        # In the case where OUs can't be mapped to a particular invocation ID (because the
+        # root plan itself might be lost), then the subinvocation_id will remain -1.
+        # diff_query_tree() will then drop the data point.
+        subinvocation_ids = subframe["subinvocation_id"].values
+        separate_subinvocation(
+            subframe["start_time"].values,
+            subframe["end_time"].values,
+            root_plans_times["start_time"].values,
+            root_plans_times["end_time"].values,
+            subinvocation_ids,
         )
+        subframe["subinvocation_id"] = subinvocation_ids
 
-        # Check for the same plan_node_id with different OU names.
-        node_ids_with_multiple_ous = [
-            int(node_id)
-            for node_id, distinct_ou_count in observed_nodes.groupby(["plan_node_id"]).size().items()
-            if distinct_ou_count > 1
-        ]
+        # Now group by subinvocation_id and apply diff_query_invocation.
+        subframe.groupby(by=["subinvocation_id"]).apply(diff_query_invocation, diffed_matrixes=diffed_matrixes)
+        return None
+
+    diff_query_invocation(subframe, diffed_matrixes=diffed_matrixes)
+    return None
+
+
+def diff_queries(unified, diffed_matrixes):
```

Comment:
A minor nitpick here and elsewhere is that I think the plural is "matrices", if you want to find and replace